src/Uno.UI/UI/Xaml/Controls/ListViewBase/ListViewBaseSource.iOSMacOS.cs

```diff
@@ -0,0 +1,518 @@
+
+		private WeakReference<NativeListViewBase> _owner;
+		public NativeListViewBase Owner
+		{
+			get { return _owner?.GetTarget(); }
+			set { _owner = new WeakReference<NativeListViewBase>(value); }
+		}
+
+
+		public ListViewBaseSource(NativeListViewBase owner)
+		{
+			Owner = owner;
+		}
+
+		#region Overrides
+#if __IOS__
+		public override nint NumberOfSections(UICollectionView collectionView)
+#else
+		public override nint GetNumberOfSections(NSCollectionView collectionView)
+#endif
+		{
+			var itemsSections = Owner.XamlParent.IsGrouping ?
+				Owner.XamlParent.NumberOfDisplayGroups :
+				1;
+			return itemsSections + SupplementarySections;
+		}
+
+#if __IOS__
+		public override nint GetItemsCount(UICollectionView collectionView, nint section)
+#else
+		public override nint GetNumberofItems(NSCollectionView collectionView, nint section)
+#endif
+		{
+			int count;
+			if (Owner.XamlParent.IsGrouping)
+			{
+				count = GetGroupedItemsCount(section);
+			}
+			else
+			{
+				count = GetUngroupedItemsCount(section);
+			}
+
+			if (this.Log().IsEnabled(Microsoft.Extensions.Logging.LogLevel.Debug))
+			{
+				this.Log().Debug($"Count requested for section {section}, returning {count}");
+			}
+			return count;
+		}
+
+		private int GetUngroupedItemsCount(nint section)
+		{
+			int count;
+			if (section == 0)
+			{
+				count = Owner.XamlParent.NumberOfItems;
+			}
+			else
+			{
+				// Extra section added to accommodate header+footer, contains no items.
+				count = 0;
+			}
+
+			return count;
+		}
+
+		private int GetGroupedItemsCount(nint section)
+		{
+			if ((int)section >= Owner.XamlParent.NumberOfDisplayGroups)
+			{
+				// Header+footer section which is empty
+				return 0;
+			}
+			return Owner.XamlParent.GetDisplayGroupCount((int)section);
+		}
+#if __IOS__
+		public override UICollectionViewCell GetCell(UICollectionView collectionView, NSIndexPath indexPath)
+#else
+		public override NSCollectionViewItem GetItem(NSCollectionView collectionView, NSIndexPath indexPath)
+#endif
+		{
+			using (
+			   _trace.WriteEventActivity(
+				   TraceProvider.ListViewBaseSource_GetCellStart,
+				   TraceProvider.ListViewBaseSource_GetCellStop
+			   )
+			)
+			{
+				// Marks this item to be materialized, so its actual measured size 
+				// is used during the calculation of the layout.
+				// This is required for paged lists, so that the layout calculation
+				// does not eagerly get all the items of the ItemsSource.
+				UpdateLastMaterializedItem(indexPath);
+
+				var index = Owner?.XamlParent?.GetIndexFromIndexPath(IndexPath.FromNSIndexPath(indexPath)) ?? -1;
+
+				var identifier = GetReusableCellIdentifier(indexPath);
+
+				var listView = (NativeListViewBase)collectionView;
+#if __IOS__
+				var cell = (ListViewBaseInternalContainer)collectionView.DequeueReusableCell(identifier, indexPath);
+#else
+				var cell = (ListViewBaseInternalContainer)collectionView.MakeItem(GetReusableCellIdentifier(indexPath), indexPath);
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Refactor this method to add validation of parameter 'collectionView' before using it.](https://app.codacy.com/gh/unoplatform/uno/pullRequest?prid=4931087)